### PR TITLE
perf: [ SPIKE ] Improved tree-shaking - DO NOT MERGE! [skip ci]

### PR DESCRIPTION
**Pull request by ze-flo**
# ⚠️ This is a SPIKE PR - DO NOT MERGE❗                                                                                                   
## Description

This PR showcases an update to `rollup.config.js` to improve tree-shaking and demonstrates improvements when applied to [@zendeskgarden/react-buttons](https://github.com/zendeskgarden/react-components/tree/main/packages/buttons) and [@zendeskgarden/react-typography](https://github.com/zendeskgarden/react-components/tree/main/packages/typography).

## Detail
- Updates `rollup.config.js` to use the [preserveModules](https://rollupjs.org/configuration-options/#output-preservemodules) option. For the ESM output, Rollup with now create separate chunks for all modules using the original module names as file names. This method improves tree-shaking and is commonly used by other popular component libraries (Ex: [Polaris](https://github.com/Shopify/polaris/blob/main/polaris-react/rollup.config.mjs#L86), [Mantine](https://github.com/mantinedev/mantine/blob/master/scripts/build/rollup/create-package-config.ts#L51), [Primer](https://github.com/primer/react/blob/main/packages/react/rollup.config.js#L243) to list a few).
- Specifies a `dist/esm` folder to include ES chunks. New folder structure:
```
.
└── dist/
    ├── esm/
    │   ├── index.js
    │   ├── elements
    │   ├── styled
    │   ├── types
    │   └── node_modules (Optional. Ex: inline svg import)
    ├── index.cjs.js
    └── typings
```
- Refines browser targets to improve audience coverage while excluding problematic and uncommon mobile browsers. This avoids bloating our bundles with unnecessary polyfills.
- Simplifies `babel.config.js` now that targeted browsers natively support the modern JS features we use.
- Drops `rollup-plugin-license` in favor of Rollup's built-in [output.banner](https://rollupjs.org/configuration-options/#output-banner-output-footer) option
- Changes the dist

## Testing the new config

To better show how a small project can better benefit from improved tree-shaking, I tested the new ES output in 2 environments while only importing `ChevronButton` from `@zendeskgarden/react-buttons` and `Paragraph` from `@zendeskgarden/react-typography`.

#### Branches
[main](https://github.com/ze-flo/shake-it-off-cra)
[tree-shaking](https://github.com/ze-flo/shake-it-off-cra/compare/main...tree-skaking-spike)

#### Results

`main`

Screen Shot 2024-04-03 at 1 00 44 PM

`@zendeskgarden`

Screen Shot 2024-04-03 at 1 06 48 PM

| **Bundle**       | **Before (KB)** | **After (KB)** | **Diff** |
|------------------|-----------------|----------------|-----------|
| main.js            | 307.02          | 212.72         | -30.71%   |
| @zendeskgarden   | 40.84           | 18.97          | -53.55%   |
| react-typography | 14.96           | 1.63           | -89.10%   |
| react-buttons    | 12.88           | 8.63           | -33.00%   |

### Webpack env

A Webpack project inspired by @colinkey's [test env](https://zendesk.atlassian.net/wiki/spaces/~617fa4bc3e3753006f618197/pages/6104220480/Tree+Shaking+Garden+Packages?focusedCommentId=6147834119).

#### Branches
[main](https://github.com/ze-flo/skake-if-off-colin)
[tree-shaking](https://github.com/ze-flo/skake-if-off-colin/compare/main...tree-shaking)

#### Results

Screen Shot 2024-04-03 at 1 55 25 PM

| **Bundle**       | **Before (KB)** | **After (KB)** | **Diff** |
|------------------|-----------------|----------------|-----------|
| main.js          | 578.76          | 384.61         | -33.55%   |
| @zendeskgarden   | 86.87           | 51.73          | -40.45%   |
| react-typography | 25.53           | 5.17           | -79.75%   |
| react-buttons    | 22.87           | 17.52          | -23.39%   |

===================================================================

- [ ] ~~:ok_hand: design updates will be Garden Designer approved (add the designer as a reviewer)~~
- [ ] ~~:globe_with_meridians: demo is up-to-date (`npm start`)~~
- [ ] ~~:arrow_left: renders as expected with reversed (RTL) direction~~
- [ ] ~~:metal: renders as expected with Bedrock CSS (`?bedrock`)~~
- [ ] ~~:guardsman: includes new unit tests. Maintain existing coverage (always >= 96%)~~
- [ ] ~~:wheelchair: tested for WCAG 2.1 AA accessibility compliance~~
- [ ] ~~:memo: tested in Chrome, Firefox, Safari, and Edge~~

### CODE REVIEW

File: babel.config.js [hunk 1]
```diff
@@ -10,19 +10,19 @@ module.exports = {
     [
       '@babel/preset-env',
       {
-        targets: '> 0.5%, last 2 versions, Firefox ESR, not dead'
+        // https://browsersl.ist/#q=%3E0.3%25%2C+last+4+versions%2C+not+dead%2C+Firefox+ESR%2C+not+and_qq+13-14%2C+not+kaios+2-5%0A
+        targets: '>0.3%, last 4 versions, not dead, Firefox ESR, not and_qq 13-14, not kaios 2-5'
```
Comment (ze-flo) [inline babel.config.js:14]: Improves the audience coverage from [88.4%](https://browsersl.ist/#q=%3E+0.5%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+dead%0A) to [89.7%](https://browsersl.ist/#q=%3E0.3%25%2C+last+4+versions%2C+not+dead%2C+Firefox+ESR%2C+not+and_qq+13-14%2C+not+kaios+2-5%0A).

File: babel.config.js [hunk 2]
```diff
@@ -10,19 +10,19 @@ module.exports = {
...
+    [
+      '@babel/preset-react',
+      {
+        useBuiltIns: true // https://babeljs.io/docs/babel-preset-react#usebuiltins
+      }
+    ], // https://babeljs.io/docs/babel-preset-react#usebuiltins
     ['@babel/preset-typescript', { onlyRemoveTypeImports: true }]
   ],
-  plugins: [
-    '@babel/plugin-transform-object-assign',
```
Comment (ze-flo) [inline babel.config.js:20]: Natively supported by targeted browsers: https://caniuse.com/mdn-javascript_builtins_object_assign

File: babel.config.js [hunk 3]
```diff
@@ -10,19 +10,19 @@ module.exports = {
...
-    ['@babel/plugin-transform-class-properties', { loose: true }],
```
Comment (ze-flo) [inline babel.config.js:21]: Natively supported by targeted browsers: https://github.com/babel/babel/blob/main/packages/babel-compat-data/data/plugins.json#L90-L102

File: babel.config.js [hunk 4]
```diff
@@ -10,19 +10,19 @@ module.exports = {
...
-    'babel-plugin-styled-components',
-    ['@babel/plugin-transform-private-property-in-object', { loose: true }],
```
Comment (ze-flo) [inline babel.config.js:23]: Natively supported by targeted browsers: https://github.com/babel/babel/blob/main/packages/babel-compat-data/data/plugins.json#L64C2-L89

File: babel.config.js [hunk 5]
```diff
@@ -10,19 +10,19 @@ module.exports = {
...
-    ['@babel/plugin-transform-private-methods', { loose: true }]
```
Comment (ze-flo) [inline babel.config.js:24]: Natively supported by targeted browsers: https://github.com/babel/babel/blob/main/packages/babel-compat-data/data/plugins.json#L116-L128

File: utils/build/rollup.config.js [hunk 6]
```diff
@@ -68,7 +74,7 @@ export default [
         babelrc: false,
         exclude: 'node_modules/**', // only transpile our source code
         ...babelOptions,
-        extensions: [...DEFAULT_EXTENSIONS, '.ts', '.tsx']
+        extensions: ['.tsx', '.ts', ...DEFAULT_EXTENSIONS]
```
Comment (ze-flo) [inline utils/build/rollup.config.js:77]: Slightly speeds up Babel compilation by targeting `.tsx?` files first.

File: utils/build/rollup.config.js [hunk 7]
```diff
@@ -99,8 +94,20 @@ export default [
...
+        file: pkg.main,
+        format: 'cjs',
+        interop: 'auto',
+        banner: BANNER // Apply global Zendesk license
+      },
+      {
+        dir: path.dirname(pkg.module),
+        format: 'es',
+        preserveModules: true,
+        preserveModulesRoot: 'src',
```
Comment (ze-flo) [inline utils/build/rollup.config.js:107]: [preserveModulesRoot](https://rollupjs.org/configuration-options/#output-preservemodulesroot) is needed to maintain the default folder structure for cases when a component imports a dependency that isn't "externalized". Ex: `ChevronButton` inlines SVG import. 

### Default `/dist` folder structure

```
.
└── dist/
    ├── esm/
    │   ├── index.js
    │   ├── elements
    │   ├── styled
    │   ├── types
    │   └── node_modules (inline svg import)
    ├── index.cjs.js
    └── typings

```

File: babel.config.js (hunk 2 above)
Comment (geotrev) [reply babel.config.js:20]: I love that you're going through these. the babel config has needed serious love. lol

**Comment by ze-flo** [comment]
I've been meticulously comparing build outputs for each component. I'm more and more confident that this upgrade may not even introduce a breaking change at all, meaning it could be released under V8. 🙌 

## TL;DR

For ex, let's compare the diff for [ColorSwatchDialog](https://www.diffchecker.com/mWmrMqaD/). 

On the left is the legacy `./dist/index.ems.js`, with every component and their relative dependencies lumped in. This was built using the current rollup config.

On the right, we have `./dist/esm/elements/ColorSwatchDialog/index.js` **only**, generated from the new rollup config.

Scroll down to `line 1421 left`, and notice an almost perfect match between the 2 outputs. The only difference is the replacement of the `_extends$2` polyfill (previously injected by `@babel/preset-react`) with the widely supported `Object.assign` (enabled via `'@babel/preset-react', { useBuiltIns: true }`)


Here's an another example for [ChevronButton](https://www.diffchecker.com/poawbGH5/). The inlined SVG is imported from `import SvgChevronDownStroke from '../node_modules/@zendeskgarden/svg-icons/src/16/chevron-down-stroke.svg.js';` (`line 9 right`). The content of this import matches the inlined icon on the left (`line 476 left`).


## Why is this good? 

It means changes to the Babel config, including [targets](https://github.com/zendeskgarden/react-components/pull/1780/files#diff-09c56b2bf95de2a608a36afef3b6893146a959d6739be0a154dc9c9f02d80f24R14) have minimal impact on the transpilation output, even when ignoring KaiOs and Android QQ browsers (their global [usage](https://browsersl.ist/#q=%3E0.3%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+dead) is 0.08% and 0.24% respectively). The compatibility between the builds is unchanged. 💯 

The reason for this positive outcome is that we don't use bleeding-edge JS features. The newest being the widely [supported](https://caniuse.com/mdn-javascript_operators_optional_chaining) optional chaining operators (ES2020).

If I've missed anything, let me know. Thx

**Review by hzhu — COMMENTED** [review]
Nice! Does this resolve https://github.com/zendeskgarden/react-components/issues/1155? 😮

File: package.json [hunk 8]
```diff
@@ -17,7 +17,7 @@
     "lint:md": "markdownlint README.md packages/*/README.md packages/*/demo/**/*.mdx",
     "new": "utils/scripts/new.mjs",
     "prepare": "npm run build",
-    "start": "storybook dev --no-version-updates -p 6006",
+    "start": "npm run build && storybook dev --no-version-updates -p 6006",
```
Comment (jzempel) [inline package.json:20]: Let's not add this overhead to the `build:demo` and `start` commands.

File: babel.config.js (hunk 2 above)
Comment (jzempel) [reply babel.config.js:20]: Big plus one. Remember to remove deleted plugins from `package.json`.

**Comment by geotrev** [comment]
Considering this doesn't seem like a breaking change, I like the idea of a V8 release too. Looking good (pending JZ comments about browser support).

File: babel.config.js (hunk 1 above)
Comment (ze-flo) [reply babel.config.js:14]: Cool. Keeping the `> 0.3%` to maintain >=80% coverage in US.


How about `> 0.3%, last 2 versions, Firefox ESR, not dead, not and_qq 13-14, not kaios 2-5`?

https://browsersl.ist/#q=%3E0.3%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+dead%2C+not+and_qq+13-14%2C+not+kaios+2-5

File: package.json (hunk 8 above)
Comment (ze-flo) [reply package.json:20]: Absolutely. This is only for this spike review in case you want check this locally.

**Comment by ze-flo** [comment]
> Nice! Does this resolve #1155? 😮

@hzhu Thanks for pointing that out. It will! 🥳

File: babel.config.js (hunk 1 above)
Comment (jzempel) [reply babel.config.js:14]: Yeah, I think that's a good target

**Comment by jzempel** [comment]
> Considering this doesn't seem like a breaking change, I like the idea of a V8 release too. Looking good (pending JZ comments about browser support).

Yep, agreed. Nice work, @ze-flo 🙌

**Comment by ze-flo** [comment]
Closing. See implementation PR #1783.
